packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,335 @@
+ *
+ * For example:
+ * - Say we have ops 1, 2, 3 and they get grouped together in a new "groupedBatch" op
+ * - Over the wire, we send just one op and the DeltaQueue will store just this single op
+ * - When other places subscribe to "push"/"op" event or call "toArray"/"peek", they will see a version that
+ * is inline with when the ops 1, 2, 3 were separate ops and will not see the "groupedBatch" op
+ */
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
+    private readonly _trackedListeners: Map<string, (...args: any[]) => void> = new Map();
+
+    constructor(
+        private readonly _queue: IDeltaQueue<T>,
+        private readonly _virtualizeValue: (value: T) => T[],
+        _storeVirtualizedSequenceNumber?: (value: T) => void,
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => {
+                        // Let VirtualDeltaManager know it should update sequence number map (for inbound)
+                        _storeVirtualizedSequenceNumber?.(task);
+
+                        // Emit event for each decomposed op as if that's how they were pushed to the DeltaQueue
+                        this._virtualizeValue(task).forEach((it) => this.emit(event, it));
+                    }
+                } else {
+                    listener = (...args: any[]) => { this.emit(event, ...args); };
+                }
+
+                this.addTrackedListener(event, listener);
+            }
+        });
+    }
+
+    private addTrackedListener(event: string, listener: (...args: any[]) => void) {
+        (this._queue as any).on(event, listener); // TODO: find better cast than "any"
+        assert(!this._trackedListeners.has(event), "double tracked listener");
+        this._trackedListeners.set(event, listener);
+    }
+
+    private removeTrackedListeners() {
+        for (const [event, listener] of this._trackedListeners.entries()) {
+            (this._queue as any).off(event, listener); // TODO: find better cast than "any"
+        }
+        this._trackedListeners.clear();
+    }
+
+    public peek(): T | undefined {
+        const value = this._queue.peek();
+        if (value === undefined) {
+            return undefined;
+        }
+        // Decompose potential "groupedBatch" and send the last message of that grouping
+        const valueArr = this._virtualizeValue(value);
+        return valueArr[valueArr.length - 1];
+    }
+
+    public toArray(): T[] {
+        const virtualizedArr: T[] = [];
+        this._queue.toArray().forEach((value) => {
```

Comment:
Is this same as this._queue.toArray().map((value) => this._virtualizeValue(value)) ?